sdb/commands/zfs/dbuf.py

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import argparse
+from typing import Iterable
+
+import drgn
+import sdb
+
+
+class DslDir():
+
+    @staticmethod
+    def name(dd):
+        pname = ''
+        if dd.dd_parent:
+            pname = DslDir.name(dd.dd_parent) + '/'
+        return pname + dd.dd_myname.string_().decode("utf-8")
+
+
+class Dataset():
+
+    @staticmethod
+    def name(ds):
+        if not ds:
+            return 'MOS'
+        name = DslDir.name(ds.ds_dir)
+        if not ds.ds_prev:
+            sn = ds.ds_snapname.string_().decode("utf-8")
+            if len(sn) == 0:
+                sn = '%UNKNOWN_SNAP_NAME%'
+            name += '@' + sn
+        return name
+
+
+class Objset():
+
+    @staticmethod
+    def name(os):
+        return Dataset.name(os.os_dsl_dataset)
+
+
+class Dbuf(sdb.Locator, sdb.PrettyPrinter):
+
+    names = ["dbuf"]
+    input_type = "dmu_buf_impl_t *"
+    output_type = "dmu_buf_impl_t *"
+
+    def _init_argparse(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            '-o',
+            '--object',
+            type=int,
+            help='filter: only dbufs of this object')
+        parser.add_argument(
+            '-l',
+            '--level',
+            type=int,
+            help='filter: only dbufs of this level')
+        parser.add_argument(
+            '-b',
+            '--blkid',
+            type=int,
+            help='filter: only dbufs of this blkid')
+        parser.add_argument(
+            '-d',
+            '--dataset',
+            type=str,
+            help='filter: only dbufs of this dataset name (or "MOS")')
+        parser.add_argument(
+            '-H',
+            '--has-holds',
+            action='store_true',
+            help='filter: only dbufs that have nonzero holds')
+
+    def pretty_print(self, dbufs):
+        print("{:>20} {:>8} {:>4} {:>8} {:>5} {}".format(
+            "addr", "object", "lvl", "blkid", "holds", "os"))
+        for dbuf in dbufs:
+            print("{:>20} {:>8d} {:>4d} {:>8d} {:>5d} {}".format(
+                hex(dbuf),
+                int(dbuf.db.db_object),
+                int(dbuf.db_level),
+                int(dbuf.db_blkid),
+                int(dbuf.db_holds.rc_count),
+                Objset.name(dbuf.db_objset)))
```

Comment:
When I run this on my system, I get entries like this:
```
  0xffff9880c8de6ca8      262    0        3     0 MOS
```
should we distinguish the MOS for which pool this entry is for? e.g. on my system I have both `domain0` and `rpool`.

Reply:
I designed this to mimic the mdb `dbuf` command, which behaves this way.   But that is bugged me for a long time so maybe now would be a good time to change it.  The canonical way to do this would be to call the dataset `poolname/$MOS` (see `MOS_DIR_NAME`), would that make sense?

Reply:
`poolname/$MOS` makes sense to me. I'm also OK with `MOS` if you are, since you'll likely be a more active user of the command than me.

> But that is bugged me for a long time so maybe now would be a good time to change it. 

I definitely think now is a great time to improve on what MDB provided us, in general. I would expect (hope) we'll be reaching audiences that won't have any prior expectations set by MDB, so we might be the only ones that really would benefit from trying to match output with MDB.

Reply:
Unfortunately the parser doesn't allow us to use `$`.  Unless someone is working on that, I'll have to find another solution.

```
> dbuf -d rpool/$MOS
usage: dbuf [-h] [-o OBJECT] [-l LEVEL] [-b BLKID] [-d DATASET] [-H]
dbuf: error: unrecognized arguments: $ MOS
```
I also tried quoting the argument with `"` or `'`, which did allow for the `$` to be passed to the `dbuf` command, but the parser didn't strip off the quotes, so that also doesn't work.

Reply:
OK.. we can leave this as-is, if you'd like to, given the constraints.

Reply:
Issued https://github.com/delphix/sdb/issues/81

Reply:
I'm going with `poolname/_MOS` for now.